lib/validation/validateNode.js

```diff
@@ -1,245 +1,35 @@
-    };
-}
-
-/**
- * A rule that ensure code lines are always children
- * of a code block.
- */
-function noOrphanLine(opts: Options) {
-    return {
-        // Match all blocks that are not code blocks
-        match(node) {
-            return (
-                (node.object === 'block' || node.object === 'document') &&
-                node.type !== opts.containerType
-            );
-        },
-
-        validate(node) {
-            const codeLines = node.nodes.filter(n => n.type === opts.lineType);
-
-            if (codeLines.isEmpty()) {
-                // All good
-                return null;
-            }
-            // Wrap the orphan lines
-            return {
-                toWrap: codeLines
-            };
-        },
-
+function validateNode(opts: Options): (node: Node) => ?Change {
+    return function validateCodeNode(node) {
         /**
-         * Wrap the given blocks in code containers
-         * @param {List<Nodes>} value.toWrap
+         * A rule that ensure code blocks contains no marks
          */
-        normalize(change, node, value) {
-            return value.toWrap.reduce(
-                (c, n) => c.wrapBlockByKey(n.key, opts.containerType),
-                change
-            );
-        }
-    };
-}
-
-/**
- * A rule that ensure code blocks only contain lines of code, and no marks
- */
-function onlyLine(opts: Options) {
-    return {
-        match: node => node.type === opts.containerType,
-
-        validate(node) {
-            const { nodes } = node;
-
-            let toAdd = List();
-            const toRemove = List();
-
-            nodes.forEach(child => {
-                if (child.object === 'text') {
-                    const lines = deserializeCode(opts, child.text).nodes;
-                    toAdd = toAdd.concat(lines);
-                    toRemove.push(child);
-                } else if (child.type !== opts.lineType) {
-                    toRemove.push(child);
-                }
-            });
-
-            if (toAdd.size || toRemove.size) {
-                return { toAdd, toRemove };
-            }
-            return null;
-        },
-
-        normalize(change, node, { toAdd, toRemove }) {
-            toAdd.forEach((child, index) => {
-                change.insertNodeByKey(node.key, index, child);
-            });
-
-            toRemove.forEach(child => {
-                change.removeNodeByKey(child.key);
-            });
-
-            // Also remove marks here (since the no mark rule for
-            // lines will not be applied afterward).
-            return applyRule(noMarks(opts), change, node.key);
-        }
-    };
-}
-
-/**
- * A rule that ensure code lines only contain one text
- * node.
- */
-function onlyText(opts: Options) {
-    return {
-        match: node => node.type === opts.lineType,
-
-        validate(node) {
-            const { nodes } = node;
+        if (!opts.allowMarks && node.type === opts.lineType) {
```

Comment:
Can't you use the marks validation from the new schema ?

https://docs.slatejs.org/other-packages/slate-schema-violations#node_mark_invalid